Webserver adding sphere tracker api

Etsuna opened the pull request:
## Summary

This merge request **adds a brand-new API module** and endpoint:

- **File added:** `WebHostLib/api/sphere_tracker.py`
- **Route:** `GET /api/sphere_tracker/<suuid:tracker>`
- **Views:**
  - **Compact (default):** IDs-only, for lightweight consumers.
  - **Expanded (`?expand=1`):** adds resolved names/games/flags derived from `TrackerData`.
- **Deterministic ordering**: sorted by `team → sphere → finder_slot → receiver_slot`, and pairs within receivers sorted by `(location_id, item_id)`.
- **Cache correctness**: memoized by `(tracker, expand)` so compact and expanded responses are cached independently.

---

## What is being added?

- An endpoint for **sphere-based progression**, grouped by team/sphere/finder/receiver.
- A **single-call expanded view** that exposes human-readable names and games while preserving the compact view for existing clients.
- Stable, predictable JSON for easier diffs, caching, and frontend rendering.

### New file(s)

```
WebHostLib/api/sphere_tracker.py
```

This module includes:

- helper collectors to compute pairs, flags, and per-game location usage;
- a compact response for legacy/low-bandwidth use;
- an expanded response (opt-in via `expand=1`) that enriches the payload with names/games/flags;
- memoized caching keyed on `(tracker, expand)`.

---

## API details

### Route

```
GET /api/sphere_tracker/<suuid:tracker>
```

### Query parameters

- `expand=1` — return the expanded payload with resolved names/games/flags.
- *(default)* — compact payload (IDs-only).

### Compact response (IDs-only)

```json
[
  {
    "team": 1,
    "spheres": [
      {
        "sphere": 2,
        "finders": [
          {
            "finder_slot": 25,
            "receivers": [
              {
                "receiver_slot": 1,
                "pairs": [[1234, 5678]]
              }
            ]
          }
        ]
      }
    ]
  }
]
```

### Expanded response (when `expand=1`)

```json
[
  {
    "team": 1,
    "spheres": [
      {
        "sphere": 2,
        "finders": [
          {
            "finder_slot": 25,
            "finder_name": "Player25",
            "finder_game": "Zillion",
            "receivers": [
              {
                "receiver_slot": 1,
                "receiver_name": "Player1",
                "items": [
                  {
                    "item_id": 1234,
                    "item_name": "13 Coins",
                    "item_flag": 0,
                    "location_id": 5678,
                    "location_name": "P-3 top far left"
                  }
                ]
              }
            ]
          }
        ]
      }
    ]
  }
]
```

> **Resolution rules**
> - `game` is carried by the **finder** (design invariant).
> - `item_name` is resolved with the **receiver's game**.
> - `location_name` is resolved with the **finder's game**.

---

## Rationale

- Frontends otherwise need secondary lookups to resolve names/games.
- The expanded view centralizes resolution server-side and preserves the compact view.
- Deterministic ordering reduces noise in PRs and stabilizes caches/UI rendering.

---

## Implementation notes

- **Cache variation on `expand`:**
  ```python
  @api_endpoints.route("/sphere_tracker/<suuid:tracker>")
  def api_sphere_tracker(tracker):
      expand = 1 if request.args.get("expand", default=0, type=int) == 1 else 0
      return _api_sphere_tracker_cached(tracker, expand)

  @cache.memoize(timeout=TRACKER_CACHE_TIMEOUT_IN_SECONDS)
  def _api_sphere_tracker_cached(tracker, expand: int):
      ...
  ```

- **Deterministic sorting:**
  - Use `sorted(...)` at each level.
  - Sort receiver pairs by `(location_id, item_id)`:
    ```python
    pairs.sort(key=lambda t: (t[1], t[0]))
    ```

- **Name resolution:**
  - `item_name` from the **receiver's game**.
  - `location_name` from the **finder's game**.
  - Player names/games from `_collect_players()`.

---

## How was this tested?

- Local server runs querying both variants:
  - `GET /api/sphere_tracker/<tracker>` (compact)
  - `GET /api/sphere_tracker/<tracker>?expand=1` (expanded)
- Verified stable ordering across multiple teams/spheres/finders/receivers.
- Confirmed caching behavior: compact/expanded responses are independent.

---

## Backward compatibility

- Adds a **new endpoint**; no changes to existing ones.
- Compact view remains IDs-only; expanded view is opt-in.

---

## Screenshots

N/A — API change only.

---

## Checklist

- [x] New endpoint (`/api/sphere_tracker/<suuid:tracker>`) added
- [x] Compact + expanded views
- [x] Deterministic ordering
- [x] Cache key includes `expand`
- [x] Tested locally

---

## Related

Add references if applicable, e.g. `Fixes #<issue-number>`.

Berserker66 commented:
I believe the expanded mode to be a mistake; it is a third cached view on the same data, one that at that point is assembled faster as html anyway and would coincide with the same cache-hit as a regular user. It also encourages not implementing proper datapackage handling.

Etsuna commented:
> I believe the expanded mode to be a mistake; it is a third cached view on the same data, one that at that point is assembled faster as html anyway and would coincide with the same cache-hit as a regular user. It also encourages not implementing proper datapackage handling.

Thanks for the review. Agreed. I removed the expanded mode so `sphere_tracker` is IDs-only. Name resolution is expected client-side via the datapackage (and aliases from `/api/tracker/<id>`). This avoids introducing a third cached view and encourages proper datapackage handling. I’ve updated the MR description to reflect this.

Etsuna commented:
I have done my changes.

Etsuna commented:
I have a little change. I'll submit it.

Etsuna commented:
Done.
